content/upcoming/enry.md

```diff
@@ -0,0 +1,183 @@
+
+## How does enry works?
+The goal of detecting language from a file is achieved by the functions:
+```
+func GetLanguage(filename string, content []byte) (language string)
+func GetLanguages(filename string, content []byte) []string
+```
+After calling `GetLanguages`, there could be more than one possible language detected for the file, so `GetLanguages` returns all of them, and `GetLanguage` returns one of them (in fact, the most likely one thanks to the classifier, we'll cover the classifier later).
+
+The detection process must be split in several steps that form a chain or sequence of strategies. This strategies need the file name, content and a list of candidates (a list of possible languages for the file). Strategies are typed functions with the next signature:
+```
+type Strategy func(filename string, content []byte, candidates []string) (languages []string)
+```
+
+Strategies try to guess the language of the file based on this arguments and a specific characteristic. For example the default strategies enry uses are:
+```
+var DefaultStrategies = []Strategy{
+	GetLanguagesByModeline,
+	GetLanguagesByFilename,
+	GetLanguagesByShebang,
+	GetLanguagesByExtension,
+	GetLanguagesByContent,
+	GetLanguagesByClassifier,
+}
+```
+You can see how the strategies look for Modeline, Filename, Shebang... characteristics of a file that can be representative of the language it uses. Some strategies needs to parse the content and perhaps apply heuristics while others, however,  do their job only with the filename.
+
+The way the strategy chain works is as follows:
+* A strategy tries to get the language from a file.
+* It can result in zero or more languages detected.
+* If there's no language, call the next strategy.
+* If there is more than one languages, they are appended to a list of candidates and they are given to the next strategy.
+* If there is exactly one language, it is returned as the detected language and the process stops.
+
+ When the detection process falls through all the strategies without getting only one language as an outcome, it reaches the last step: `GetLanguagesByClassifier`. This strategy deserves taking a look at it.
+
+`GetLanguagesByClassifier` internally uses an object of the type:
+```
+type Classifier interface {
+	Classify(content []byte, candidates map[string]float64) (languages []string)
+}
+```
+
+This interface allows you to implement your own classifier and use it in the following functions:
+```
+func GetLanguageBySpecificClassifier(content []byte, candidates []string, classifier Classifier) (language string, safe bool)
+func GetLanguagesBySpecificClassifier(content []byte, candidates []string, classifier Classifier) (languages []string)
+```
+
+*enry*'s default classifier implementation is a bayesian classifier which is like the one *linguist* uses. It assigns scores to the candidates regarding to its probability, with the highest score assigned to the most likely candidate.
```

Comment:
Here could be a good chance to add a link to Wikipedia for details about the family of classification algorithms being used.